src/silo/query_engine/actions/aggregated.cpp

```diff
@@ -0,0 +1,304 @@
+         is_string = true;
+         index = string_index - 1;
+      }
+   }
+   return TupleAccessor{index, is_string, ascending};
+}
+
+void applyOrderBy(std::vector<Tuple>& result, const std::vector<TupleAccessor>& order_by_indexes) {
+   auto cmp = [&order_by_indexes](const Tuple& value1, const Tuple& value2) {
+      for (const auto& accessor : order_by_indexes) {
+         if (accessor.is_string) {
+            if (value1.string_columns[accessor.index] < value2.string_columns[accessor.index]) {
+               return accessor.ascending;
+            }
+            if (value2.string_columns[accessor.index] < value1.string_columns[accessor.index]) {
+               return !accessor.ascending;
+            }
+         } else {
+            if (value1.int_columns[accessor.index] < value2.int_columns[accessor.index]) {
+               return accessor.ascending;
+            }
+            if (value2.int_columns[accessor.index] < value1.int_columns[accessor.index]) {
+               return !accessor.ascending;
+            }
+         }
+      }
+      return false;
+   };
+   std::sort(result.begin(), result.end(), cmp);
+}
+
+std::vector<AggregationResult> generateResult(
+   std::vector<Tuple>& tuples,
+   std::optional<uint32_t> limit,
+   const std::vector<config::DatabaseMetadata>& field_defs
+) {
+   std::vector<AggregationResult> result;
+   const uint32_t result_size = limit.has_value()
+                                   ? std::min(limit.value(), static_cast<uint32_t>(tuples.size()))
+                                   : tuples.size();
+   for (size_t idx = 0; idx < result_size; idx++) {
+      std::map<std::string, std::variant<std::string, uint32_t>> fields;
+      uint32_t int_index = 0;
+      uint32_t string_index = 0;
+      for (const auto& field : field_defs) {
+         if (field.type == config::ValueType::NUMBER) {
+            fields[field.name] = tuples[idx].int_columns[int_index++];
+         } else if (field.type == config::ValueType::DATE) {
+            fields[field.name] = silo::common::dateToString(tuples[idx].int_columns[int_index++]);
+         } else {
+            fields[field.name] = tuples[idx].string_columns[string_index++];
+         }
+      }
+      result.push_back({fields});
+   }
+   return result;
+}
+
+}  // namespace silo::query_engine::actions
+
+template <>
+struct std::hash<silo::query_engine::actions::Tuple> {
+   std::size_t operator()(const silo::query_engine::actions::Tuple& tuple) const {
+      std::size_t seed = tuple.string_columns.size();
+      for (const std::string& value : tuple.string_columns) {
+         auto hash = std::hash<std::string>()(value);
+         // NOLINTBEGIN(readability-magic-numbers)
+         hash = ((hash >> 16) ^ hash) * 0x45d9f3b;
+         hash = ((hash >> 16) ^ hash) * 0x45d9f3b;
+         hash = (hash >> 16) ^ hash;
+         seed ^= hash + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+         // NOLINTEND(readability-magic-numbers)
+      }
+      return seed;
```

Comment:
I don't know much about hash functions, but
* why is this so complicated?
* why does this not involve `int_columns`?

From a naive point of view I would have expected that this function looks somewhat like

```
return hash(string_columns) + hash(int_columns);
```

Reply:
But why do we need to implement vector hashing ourselves? Isn't this already implemented in the standard lib?

Reply:
Not really, I am sure there is a boost lib for that, but this is more performant. I refactored it and push the changes again. Maybe this makes the whole hashing topic more self-contained